README.md
[![GitHub license](https://img.shields.io/github/license/podput/mc-data?style=for-the-badge)](https://github.com/podput/mc-data/blob/main/LICENSE)  ![GitHub last commit](https://img.shields.io/github/last-commit/podput/mc-data?style=for-the-badge)  ![GitHub top language](https://img.shields.io/github/languages/top/podput/mc-data?style=for-the-badge)

# Minecraft Stats Analyzer 

![Minecraft](https://icon-library.com/images/minecraft-logo-icon/minecraft-logo-icon-21.jpg)  

Tool to create graphs and analyze Minecraft stats

**See a quick demo of the cli application part of this project [over here](https://asciinema.org/a/447891)**   
*Along with the charts:*

![Bar Chart](./history/media/all_dropped_items.png)
![Pie Chart](./history/media/All_mined_blocks.png)  


--------------------------------------------------
**With this project, I'll experiment a little, and upload every single idea, plan, schedule, etc. associated with said project.**  
[You can see that over here](./history/history.md)


history/history.md
1. [First notes - 30.10.2021](./notes/01_Starting_ideas.pdf)
2. [Chart ideas - 9.11.2021](./notes/02_Chart_ideas.pdf)


mc_data/main.py
from Stats import *
import numpy as np
import matplotlib.pyplot as plt


def top(
    chart_type:str="bar",
    data=Stats().get_data(data_type="minecraft:used"),
    limit:int=10,
    chart_only:bool=True,
    filename:str="chart.png"):
    """Saves a graph of given data, regarding top n number of data type given.
    
    `e.g. top 12 of data type 'minecraft:broken'`
    
    Parameters
    ----------

    chart_type : Chart type. Right now you can only choose between 'pie' and 'bar'.

    data : defaults to `Stats().get_data(data_type="minecraft:used")`, but you can use it with your own instance of Stats() with your own data_type
    
    limit : defaults to `10`, you can put any integer there, but the bigger it is, the worse the chart will look.

    chart_only : is `True` by default. If set to `False`, you'll get the top n number of the data type given in a dict.

    filename : is `chart.png` by default. It's the filename of the chart.

    Returns
    -------

    dict_new : if the `chart_only` flag is set to False

    otherwise returns `True` if everything happens without error. 
    """
    
    all_data = {}

    for i in data:
        for l in i:
            for le in i[l]:
                # print(le)
                # all_data.update(le=i[l][e])
                try:
                    all_data[le] += i[l][le]
                except KeyError:
                    all_data[le] = i[l][le]
        
    
    dict_new = dict(sorted(all_data.items(), key=lambda x:x[1], reverse=True)[:limit])
    for i in list(dict_new):
        dict_new[i.replace("minecraft:", "")] = dict_new[i]
        del dict_new[i]
    
    if not chart_only:
        return dict_new

    if chart_type == "bar":
        plt.figure(figsize=(20, 10))
        plt.bar(range(len(dict_new)), list(dict_new.values()), align='center', width=0.3)
        plt.xticks(range(len(dict_new)), list(dict_new.keys()))

    elif chart_type == "pie":
        y = np.array(list(dict_new.values()))
        mylabels = list(dict_new.keys())

        plt.pie(y, labels = mylabels, autopct='%1.1f%%')
        
    else:
        raise TypeError("This only supports either 'bar' or 'pie' charts as of now.")
    
    plt.title(filename)
    plt.savefig(filename, bbox_inches='tight', dpi=200)
    print(f"Saved graph as {filename}!")
    
    return True
    # plt.rcParams["figure.figsize"] = [7.50, 3.50]
    # plt.rcParams["figure.autolayout"] = True






mc_data/Stats.py
import json
import os
import requests
import logging 
   
logger_format = '%(asctime)s:%(threadName)s:%(message)s'
logging.basicConfig(format=logger_format, level=logging.INFO, datefmt="%H:%M:%S")


class Stats():
    """Gets all/user data from files
       
        Parameters
        ----------

        path : Specific path of the 'stats' folder, defaults to '../stats'.
 
        user : Name of user, defaults to None => all users.

        Returns
        -------

        player_data : dict with the player(s) data.   
        
        player_names : list with the player name(s).

    """
    def __init__(self,
    path=os.path.abspath("../stats"),
    user:str=None,
    verbose:bool=False):

        self.verbose = verbose
        try:
            for i in os.listdir(path):
                if "-" in i:
                    new_name = i.replace("-", "")
                    os.rename("{0}/{1}".format(path, i), "{0}/{1}".format(path, new_name))
                    if verbose:
                        logging.info(f"File format of {i} corrected.")        

        except Exception as e:
            if verbose:
                logging.info(e)
                logging.info("This usually means that your files are already corrected.")
            pass

        if user:
            uuid = requests.get(f"https://api.minetools.eu/uuid/{user}").json()["id"]
            with open("{0}/{1}.json".format(path, uuid), "r") as f:
                player = json.load(f)
            self.player_data = {user:player}
            self.player_names = user
        
        else:
            player_data, player_names = {}, []
        
            for i in os.listdir(path):

                if ".json" in i:
                    with open(f"{path}/{i}", "r") as f:
                        
                        player = json.load(f)
                        player_name = i.strip(".json")
                        player_name = requests.get(f"https://api.minetools.eu/uuid/{player_name}").json()["name"]
                        player_names.append(player_name)
                        player_data[f"{player_name}"] = player
            
                else:
                    if verbose:
                        logging.info(f"Non json file - {i}\nSkipping...")
                    pass

            self.player_data = player_data
            self.player_names = player_names

    def get_data(
        self,
        data_type:str=None,
        data:dict=None) -> list:
        """Gets specific data for the player(s) given
        
        Parameters
        ----------

        data : player data list with json/dict, defaults to None, which leads to using the data for the user(s) given when initializing the class.
        
        data_type : string which is a choice between the different types of data contained in the stats files.

        Returns
        -------

        mined_blocks : dict with all the mined blocks and their frequency
        """
        if not data:
            data = self.player_data


        data_types = []        
        for i in data:
            for l in data[i]["stats"]:
                if l not in data_types:
                    data_types.append(l)


        if not data_type or data_type not in data_types:
            raise TypeError("\nPlease choose one of these as the data_type parameter:\n" + str(data_types))


        end_result = []

        for i in data:
            if data_type in data[i]["stats"]:
                end_result.append({i: data[i]["stats"][data_type]})
            else:
                if self.verbose:
                    logging.info(f"{i} hasn't {data_type}")
                pass

        return end_result


src/main_classes.py
import json
import os
import requests
import logging 
   
logger_format = '%(asctime)s:%(threadName)s:%(message)s'
logging.basicConfig(format=logger_format, level=logging.INFO, datefmt="%H:%M:%S")


class Stats():
    """Gets all/user data from files
       
        Parameters
        ----------
        ```
        path: Specific path of the 'stats' folder, defaults to '../stats'.
 
        user: Name of user, defaults to None => all users.
        ```
        Returns
        -------
        ```
        player_data: dict with the player(s) data.   
        
        player_names: list with the player name(s).
        ```

    """
    def __init__(self, path=os.path.abspath("../stats"), user=None):

        try:
            for i in os.listdir(path):
                if "-" in i:
                    new_name = i.replace("-", "")
                    os.rename("{0}/{1}".format(path, i), "{0}/{1}".format(path, new_name))
                    logging.info(f"File format of {i} corrected.")        

        except Exception as e:
            logging.info(e)
            logging.info("This usually means that your files are already corrected.")


        if user:
            uuid = requests.get(f"https://api.minetools.eu/uuid/{user}").json()["id"]
            with open("{0}/{1}.json".format(path, uuid), "r") as f:
                player = json.load(f)
            self.player_data = {user:player}
            self.player_names = user
        
        else:
            player_data, player_names = {}, []
        
            for i in os.listdir(path):

                if ".json" in i:
                    with open(f"{path}/{i}", "r") as f:
                        
                        player = json.load(f)
                        player_name = i.strip(".json")
                        player_name = requests.get(f"https://api.minetools.eu/uuid/{player_name}").json()["name"]
                        player_names.append(player_name)
                        player_data[f"{player_name}"] = player
            
                else:
                    print(f"Non json file - {i}\nSkipping...")

            self.player_data = player_data
            self.player_names = player_names

    def mined(self, data=None):
        """Gets all blocks mined for the player(s) given
        
        Parameters
        ----------

        data : player data list with json/dict, defaults to None, which leads to using the data for the user(s) given when initializing the class.
        
        Returns
        -------

        mined_blocks : dict with all the mined blocks and their frequency
        """

        if not data:
            data = self.player_data
        mined_blocks = []


        for i in data:
            mined_blocks.append({i: data[i]["stats"]["minecraft:mined"]})
        
        return mined_blocks



a = Stats=()

mined = a.mined()
print(mined)

mc_data/mc_data.py
import Stats
import numpy as np
import matplotlib.pyplot as plt


def top(
    chart_type:str="bar",
    data=Stats.Stats().get_data(data_type="minecraft:used"),
    limit:int=10,
    chart_only:bool=True,
    filename:str="chart",
    blitem:str=None):
    """Saves a graph of given data, regarding top n number of data type given.
    
    `e.g. top 12 of data type 'minecraft:broken'`
    
    Parameters
    ----------

    chart_type : Chart type. Right now you can only choose between 'pie' and 'bar'.

    data : defaults to `Stats().get_data(data_type="minecraft:used")`, but you can use it with your own instance of Stats() with your own data_type
    
    limit : defaults to `10`, you can put any integer there, but the bigger it is, the more data it will collect (the worse the chart will look).

    chart_only : is `True` by default. If set to `False`, you'll get the top n number of the data type given in a dict.

    filename : is `chart.png` by default. It's the filename/name of the chart.

    blitem :  Makes a leaderboard of a category --> block/item of choice (in category set within Stats instance)

    Returns
    -------

    dict_new : if the `chart_only` flag is set to False

    otherwise returns `True` if everything happens without error. 
    """
    
    all_data = {}


    if not blitem:
        for i in data:
                for l in i:
                    for le in i[l]:
                        # print(le)
                        # all_data.update(le=i[l][e])
                        try:
                            all_data[le] += i[l][le]
                        except KeyError:
                            all_data[le] = i[l][le]
   
    else:
        all_data = {}
        for i in data:
            for l in i:
                for le in i[l]:
                    if le == blitem:
                        all_data[l] = i[l][le]
        if not all_data:
            raise KeyError("No such block/item in this category/dataset")
    
    dict_new = dict(sorted(all_data.items(), key=lambda x:x[1], reverse=True)[:limit])
    for i in list(dict_new):
        dict_new[i.replace("minecraft:", "")] = dict_new[i]


    if not chart_only:
        return dict_new

    if chart_type == "bar":
        plt.figure(figsize=(20, 10))
        plt.bar(range(len(dict_new)), list(dict_new.values()), align='center', width=0.3)
        plt.xticks(range(len(dict_new)), list(dict_new.keys()))

    elif chart_type == "pie":
        y = np.array(list(dict_new.values()))
        mylabels = list(dict_new.keys())

        plt.pie(y, labels = mylabels, autopct='%1.1f%%')
        
    else:
        raise TypeError("This only supports either 'bar' or 'pie' charts as of now.")
    
    plt.title(filename)
    if " " in filename:
        filename = filename.replace(" ", "_")

    plt.savefig(f"{filename}.png", bbox_inches='tight', dpi=200)
    print(f"Saved graph as {filename}.png!")
    
    return True
    # plt.rcParams["figure.figsize"] = [7.50, 3.50]
    # plt.rcParams["figure.autolayout"] = True




if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(prog="python3 mc_data", description="CLI interface of mc_data")
    group = parser.add_mutually_exclusive_group()
    group.add_argument("--bar", "-b", dest="ch_type", action="store_const", const="bar", help="Generate a bar chart from top n results")
    group.add_argument("--pie", "-pie", dest="ch_type", action="store_const", const="pie", help="Generate a pie chart from top n results")
    parser.add_argument("--limit", "-l", default=10, type=int, help="Set a limit to the data put in the chart")
    parser.add_argument("--name", "-n", required=True, type=str, help="Set (file)name of the chart")
    parser.add_argument("--user", "-u", default=None, type=str, help="Get chart for a specific user")   
    parser.add_argument("--category", "-c", required=True, type=str, help="Choose what category of data to use for the chart")
    parser.add_argument("--verbose", "-v", action="store_true", default=False,  help="Enable verbose mode")
    parser.add_argument("--blitem", "-bi", default=None, type=str, help="Choose what block/item to use for the chart")
    parser.set_defaults(ch_type="bar")

    args = parser.parse_args()
    
    data_user = Stats.Stats(verbose=args.verbose, user=args.user).get_data(data_type=f"{args.category}")
   
    top(chart_type=args.ch_type, limit=args.limit, data=data_user, filename=args.name, blitem=args.blitem)

